### PR TITLE
Topic/expose graphql source query

### PR DESCRIPTION
**Pull request by MarcMcIntosh**
## Description

One of the plugins used by `gatsby-source-prismic-graphql`, `gatsby-source-graphql-universal` overrides `babel-plugin-remove-graphql-queries.js` so that the queries can be retained within `gatsby-node.js#createPages` can passed to components and re-executed at run-time.

At prismic we re-use graphql queries for previews, but this mechanism for this relies on a "on-the-fly" replacement of the gatsby `babel-plugin-remove-graphql-queries`. Thus we would like to add this feature to gatsby :)

## Related Issues
Addresses #22919

### CODE REVIEW

**Comment by MarcMcIntosh** [comment]
@pieh thanks for assigning yourself :)

The circleci tests are failing at `validate-starters.sh` the areas which it fails seem unrelated to this PR but can be fixed with.

```diff
diff --git a/scripts/validate-starters.sh b/scripts/validate-starters.sh
index 052e4c40f..1e5da7421 100644
--- a/scripts/validate-starters.sh
+++ b/scripts/validate-starters.sh
@@ -13,12 +13,13 @@ for folder in $GLOB; do

   # validate
   cd "$folder" || exit
-  npm audit &&
+  npm i --package-lock-only &&
+  npm audit fix &&
   npm install &&
   # check both npm and yarn, sometimes yarn registry lags behind
   rm -rf node_modules &&
   yarn &&
-  npm run build ||
+  [[ $(basename $folder) != "gatsby-starter-plugin" ]] && npm run build ||
   exit 1

   cd "$BASE" || exit
```

I'm not able to see the results of the google cloud tests, so more information on those would be appreciated :)

**Comment by pieh** [comment]
Please ignore the `starters_validate` - we don't have it as required check, because of some problems that were out of our direct control (transitive dependencies)

**Comment by pieh** [comment]
Will be checking the changes right now to understand changes a bit more

**Comment by pieh** [comment]
One of the things I mentioned in the issue is that the query text should not be kept in the production code if it's not referenced by anything other than Gatsby components/hooks.

For `useStaticQuery` with the current code it works well, because we replace entirety of `useStaticQuery` call in production code with imported .json file. 

### `useStaticQuery` inline way ✅:
```jsx
const data = useStaticQuery(graphql` <some_query `)
```

this gets replaced with 
```jsx
import staticQueryData from "<path_to_json_file>"
[...]
const data = staticQueryData
```

so just code tranformation removes the transformed `graphql` template literal - we are good here

### `useStaticQuery` referenced query ✅:

```jsx
const query = graphql` <some_query> `
const data = useStaticQuery(query)
```

turns into 

```jsx
import staticQueryData from "<path_to_json_file>"

const query = { /* new object structure */ }
const data = staticQueryData
```

The `query` will be removed if it's not referenced by anything - which is exactly what we want. Just note that our babel transformation here doesn't handle removal - we rely on minifyjs/terser here

### `<StaticQuery>` (both inline and referenced query) ❌ :

```jsx
<StaticQuery query={graphql` <some_query> `} render={data => { /* return some component */} />
```

turns into:

```jsx
import staticQueryData from "<path_to_json_file>"
[...]
<StaticQuery query={{ /* new object structure */ }} render={data => { /* return some component */} data={staticQueryData} />
```

Note that this wasn't introduced in this PR - in `master` we are passing query hash as query prop. But this not really needed in production, because we pass `data` prop which makes the `query` prop unused but webpack will not know that and will keep it in code. We still need query to be passed in development for data hot-reload reasons (we need to get hash there to identify query).

So I think what we would have to additionally handle here is to remove `query` prop in production for `<StaticQuery>` component case. As this is wasteful in master already - this could be split to separate PR to not overload this one.

Also ping @jfrolich to check if this change (changing from string literal (query hash) to object) would cause problems with Reason or not. I vaguely remember that some of the tooling needed some value to be present, but don't remember details.

---

Other than this - I wonder if different avenue wouldn't be just adding support for something like this:

```js
const queryText = ` <some_query>`
const query = graphql(queryText)
```

so folks could reference `queryText` directly in runtime queries (instead of `query.source`). Upside - we don't change internal structure of what we replace ``graphql`<query>` `` with, downside - possibly less nice integration with live previews?

File: packages/babel-plugin-remove-graphql-queries/src/index.js [hunk 1]
```diff
@@ -168,6 +168,23 @@ function isUseStaticQuery(path) {
   )
 }
 
+function getGraphqlExpression(t, queryHash, source) {
```
Comment (pieh) [inline packages/babel-plugin-remove-graphql-queries/src/index.js:171]: We are creating object here, so `get` "operation" threw me a bit at first. Could this have more descriptive naming? Maybe something just replace `get` with `create` or `generate`.

Also some description what this is doing (jsdoc) would be nice.

(This is preemptive comment so I don't forget and doesn't need to be addressed quickly - we still want to evaluate if replacing current query hash with object is the way to go)

**Comment by jfrolich** [comment]
> Also ping @jfrolich to check if this change (changing from string literal (query hash) to object) would cause problems with Reason or not. I vaguely remember that some of the tooling needed some value to be present, but don't remember details.

No it doesn't matter for the reason implementation.

**Comment by jfrolich** [comment]
> Other than this - I wonder if different avenue wouldn't be just adding support for something like this:
> 
> ```js
> const queryText = ` <some_query>`
> const query = graphql(queryText)
> ```
> 
> so folks could reference `queryText` directly in runtime queries (instead of `query.source`). Upside - we don't change internal structure of what we replace `` graphql`<query>` `` with, downside - possibly less nice integration with live previews?

Implementing this won't be easy as probably a lot of code needs to change to allow this as an alternative to the tagged template literal (I brought this up as a way to achieve Reason interop). But this can definitely work. As a small extra benefit Reason interop becomes much easier as Bucklescript doesn't have support for the tagged template literal without tricks.

File: packages/babel-plugin-remove-graphql-queries/src/index.js (hunk 1 above)
Comment (MarcMcIntosh) [reply packages/babel-plugin-remove-graphql-queries/src/index.js]: Hello @pieh,
I've made the changes and added some jsdoc style code-comments in commit 
[11a6f77](https://github.com/gatsbyjs/gatsby/pull/23423/commits/11a6f774fa28c42c605c6b955d8baa2dbc63120a)

**Comment by MarcMcIntosh** [comment]
> Other than this - I wonder if different avenue wouldn't be just adding support for something like this:
```js
const queryText = ` <some_query>`
const query = graphql(queryText)
```
so folks could reference queryText directly in runtime queries (instead of query.source). Upside - we don't change internal structure of what we replace graphql`<query>` with, downside - possibly less nice integration with live previews?

Do you mean passing queryText as a prop to a HOC?... that's an interesting idea :)


>So I think what we would have to additionally handle here is to remove query prop in production for <StaticQuery> component case. As this is wasteful in master already - this could be split to separate PR to not overload this one.

Sounds good :)

@pieh  Would you like anything else done on this PR?

**Comment by pieh** [comment]
For now I asked few more folks on the team about their opinions, will post update when I get it.

I do want to mention few things about `gatsby-source-graphql` (and similar plugins like `gatsby-source-graphql-universal`): this way of hooking data into Gatsby works nicely as "plug-and-play" way without creating dedicated source plugins, but it has a lot of limitations.

This approach is very problematic when your site get big enough as we are not able to determine which pages will change ahead of time to only build parts of site ( https://www.gatsbyjs.org/docs/page-build-optimizations-for-incremental-data-changes/ / incremental builds on Gatsby Cloud ). Depending on your site setup it might or might not matter - if you only ever query once for particular data on your site - it's fine (because source sync would do the same unless it supports delta sync) - but if you query it multiple times - you will always hit graphql server instead of local cache meaning that this approach scale much worse than source plugins.

It also doesn't play nice with transformer plugins ecosystem of Gatsby. 

We don't plan on removing it, but it's also not something we want to recommend.

Above is not to say we will not move forward with the PR - that's more about communicating the drawbacks (from our standpoint) of schema stitching approach. It is still super useful to get up and running quickly if you have graphql server already if you are willing to accept limitations (or those limitations don't apply to your case)

While still waiting for the feedback I do want to say that `<StaticQuery>` case will block the PR. When I mentioned that it doesn't have to happen in this PR - I meant that it's not direct problem of this PR, but is still problem we need to solve.

But maybe instead of removing prop. We could change it to use just literal hash value (not via `query.id` / `query.toString()`)?

**Comment by MarcMcIntosh** [comment]
HI @pieh 

>But maybe instead of removing prop. We could change it to use just literal hash value (not via query.id / query.toString())?

Do you mean something like this?
```js
const id = "2626356014"
const queryText = "{ query: ... }"
const query = { [id]: queryText }
```

> This approach is very problematic when your site get big enough as we are not able to determine which pages will change ahead of time to only build parts of site ( https://www.gatsbyjs.org/docs/page-build-optimizations-for-incremental-data-changes/ / incremental builds on Gatsby Cloud ). Depending on your site setup it might or might not matter - if you only ever query once for particular data on your site - it's fine (because source sync would do the same unless it supports delta sync) - but if you query it multiple times - you will always hit graphql server instead of local cache meaning that this approach scale much worse than source plugins.

I think i understand the issue with source plugins, in our use case if the content editor is using prismic previews we need to re-run the query with a different ref header as previews are handled client side. So we would probably hit the cache. 

Can you provide more information on incremental builds, I'm thinking of a mechanism similar to [make](https://www.gnu.org/software/make/)?

>While still waiting for the feedback I do want to say that <StaticQuery> case will block the PR. When I mentioned that it doesn't have to happen in this PR - I meant that it's not direct problem of this PR, but is still problem we need to solve.

Is any one working on this?

**Comment by pieh** [comment]
> Do you mean something like this?
> 
> ```js
> const id = "2626356014"
> const queryText = "{ query: ... }"
> const query = { [id]: queryText }
> ```

Currently I was refering to `<StaticQuery>` component and `query` prop:

Input: 
```js
const query = graphql` <some_query>`
<StaticQuery query={query} render={data => { /* return some component */} />
```

Output:

```js
const query = { 
  id: `abcdef`,
  source: `<some_query>`
}

<StaticQuery query={`abcdef`} render={data => { /* return some component */} />
```
so the prop is not same thing it was - just the prop get replaced with literal value.

But we will also be changing `useStaticQuery` as well (so it's not actually removed like it is right now) so it will need similar handling. StaticQuery results right now (both `<StaticQuery>` component and `useStaticQuery` hook) are problematic for incremental builds / GATSBY_EXPERIMENTAL_PAGE_BUILD_ON_DATA_CHANGES - because changes in content cause changes in webpack hashes (as static query results are right now bundled by webpack) and because of that, it forces gatsby to rebuild every html page to update `.js` file hashes. 

You can check just opened draft PR ( https://github.com/gatsbyjs/gatsby/pull/23837 - it's very far from being mergeable - but it works already for production builds ). It no longer removes whole `useStaticQuery` (and with it we no longer replace `useStaticQuery` completely with results, and instead we keep the call, so we would need similar handling as prop in `<StaticQuery>` component).

> I think i understand the issue with source plugins, in our use case if the content editor is using prismic previews we need to re-run the query with a different ref header as previews are handled client side. So we would probably hit the cache.
> 
> Can you provide more information on incremental builds, I'm thinking of a mechanism similar to [make](https://www.gnu.org/software/make/)?

When you use non-stitched schema that comes from data-layer that is "owned" by Gatsby - Gatsby knows which objects are used for what page. Because gatsby also track if objects change we selectively invalidate only query results which data source have changed - it's very fine grained. Not sure if it will help but you might get high level overview of this mechanism in https://www.gatsbyjs.org/docs/page-node-dependencies/ . Now stiched schema is "black box" for gatsby - which means we have to rerun every query against real server on every build because we have no idea if data changed or not.

"But why it matters if gatsby need to put data in data layer first?". In a lot of cases this is correct and we do full data sync during `source and transform` step - but there are sources that you can ask for data that changed since last sync (what I referred to as "delta sync" in previous comment) - in that case cached data is just hydrated and only things that changed, were added or deleted are adjusted in our data layer.

Things like this enable us to not even run queries because we know that result won't change.

> Is anybody working on this issue?

Not ATM

**Comment by MarcMcIntosh** [comment]
Hi @pieh 

We'll keep an eye on #23837 :)

>But we will also be changing useStaticQuery as well (so it's not actually removed like it is right now) so it will need similar handling. StaticQuery results right now (both <StaticQuery> component and useStaticQuery hook) are problematic for incremental builds / GATSBY_EXPERIMENTAL_PAGE_BUILD_ON_DATA_CHANGES - because changes in content cause changes in webpack hashes (as static query results are right now bundled by webpack) and because of that, it forces gatsby to rebuild every html page to update .js file hashes.


 I've taken a look through the PR #23837. `staticQueryDb` seems like an interesting mechanism it's using the pagePath as the key in the hash-map and the query results as the value. 

It makes me wonder if the data or queries for each pages are combined? As most of the time I've used static queries it has been at the component level, for say a header / footer and navigation that will be used in layouts (multiple pages)


>Currently I was refering to <StaticQuery> component and query prop:
Input:
```jsx
const query = graphql` <some_query>`
<StaticQuery query={query} render={data => { /* return some component */} />
```
Output:
```jsx
const query = { 
  id: `abcdef`,
  source: `<some_query>`
}
```
```jsx
<StaticQuery query={`abcdef`} render={data => { /* return some component */} />
```
so the prop is not same thing it was - just the prop get replaced with literal value.


With that we we could write a test for the changes to `<StaticQuery>` :sweat_smile:
For what I can tell PR #23837 it looks more like changes to the build process?

**Comment by pieh** [comment]
> It makes me wonder if the data or queries for each pages are combined? As most of the time I've used static queries it has been at the component level, for say a header / footer and navigation that will be used in layouts (multiple pages)

Data will not be combined in `page-data` files - but page-data files will have just list of static queries it needs to additionally fetch (static query results will be separate files), so there won't be pure data "bloat" in page-data files.

**Comment by pieh** [comment]
> With that we we could write a test for the changes to `<StaticQuery>` 😅

Yeah we should - we already have those snapshot tests that would show babel transform changes - particularly to cases where you define query as separate variable. 

Sidenote - having so many snapshots in separate file makes it very hard to follow those - I might consider changing our assertions from `toMatchSnapshot` to `toMatchInlineSnapshot` - it will make the test file unwieldy (because it would be much bigger than it is today) ... but the results will be collocated to tests so it's much easier to track which snapshot belong to which test

> For what I can tell PR #23837 it looks more like changes to the build process?

Well, yes - but it relies on not removing `useStaticQuery` fully in production `build-javascript` step (right now - later on we will also want same behaviour for `build-html` which still removes it in current shape of draft PR).

We also do have a problem with our jest tests today because we don't exactly check for those various scenarios and code paths (It least I think we don't) - like we have checks like
```
[`production`, `test`].includes(process.env.NODE_ENV)
```
which will be always true in jest tests I think because `NODE_ENV` is `test`

**Comment by pieh** [comment]
I'll put this PR in draft mode for now (as there are few things that are not handled here)

**Comment by MarcMcIntosh** [comment]
Hi @pieh, yes it's a good idea to put this in draft until #23837 is complete :)

**Comment by sanscheese** [comment]
#23873 linked to https://github.com/gatsbyjs/gatsby/pull/25723 which is now merged. Does that mean this can possibly be ready to go?

**Comment by pragmaticpat** [comment]
@MarcMcIntosh @vladar @pieh - is this still valid?

**Comment by MarcMcIntosh** [comment]
Hi @pragmaticpat, we've stopped supporting the  `gatsby-source-prismic-graphql` and `gatsby-source-graphql-universal` plugins infavour of `gatsby-source-prismic`.

I know we did (possibly still do) have limitations with previews in `gatsby-source-prismic` that could be improved if we could rerun queries client side. I'll tag @angeloashmore he's the best person to talk with about this :)

**Comment by LekoArts** [comment]
I'll close this for now as it went stale and the initial plugin which wanted this was switched out to a different plugin.
